src/provider/meta.ts

```diff
@@ -0,0 +1,63 @@
+import type { ExternalProvider, JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+import WalletConnectProvider from '@walletconnect/ethereum-provider'
+
+function isWeb3Provider(provider: JsonRpcProvider): provider is Web3Provider {
+  return 'provider' in provider
+}
+
+function isWalletConnectProvider(provider: ExternalProvider): provider is WalletConnectProvider {
+  return (provider as WalletConnectProvider).isWalletConnect
+}
+
+export enum WalletType {
+  WALLET_CONNECT = 'WalletConnect',
+  INJECTED = 'Injected',
+  UNKNOWN = 'Unknown',
+}
+
+export interface WalletMeta extends Partial<NonNullable<WalletConnectProvider['connector']['peerMeta']>> {
+  type: WalletType
```

Comment:
I think I would make it a bit clearer here that `peerMeta` can be either WalletConnect `peerMeta` or an equivalent for injected that we construct ourselves.

Maybe something like this: `type WalletMeta = { type: WalletType, meta: WalletConnectPeerMeta | InjectedProviderPeerMeta }`

Reply:
My intent was to have an abstract type. I'll type it directly instead of deriving it.